src/rest-server/src/middlewares/v2/protocol.js

```diff
@@ -0,0 +1,162 @@
+  'script',
+  'output',
+  'data',
+  'dockerImage',
+];
+
+const protocolWrap = async (req, res, next) => {
+  const protocolYAML = req.body;
+  req.rawBody = protocolYAML;
+  req.body = {
+    protocol: protocolYAML,
+  };
+  await next();
+};
+
+const protocolValidate = async (req, res, next) => {
+  const protocolYAML = req.body.protocol;
+  const protocolJSON = yaml.safeLoad(protocolYAML);
+  if (!protocolSchema.validate(protocolJSON)) {
+    throw createError('Bad Request', 'InvalidProtocolError', protocolSchema.validate.errors);
+  }
+  // convert prerequisites list to dict
+  const prerequisites = {};
+  for (let type of prerequisiteTypes) {
+    prerequisites[type] = {};
+  }
+  if ('prerequisites' in protocolJSON) {
+    for (let item of protocolJSON.prerequisites) {
+      prerequisites[item.type][item.name] = item;
+    }
+  }
+  protocolJSON.prerequisites = prerequisites;
+  // convert deployments list to dict
+  const deployments = {};
+  if ('deployments' in protocolJSON) {
+    for (let item of protocolJSON.deployments) {
+      deployments[item.name] = item;
+    }
+  }
+  protocolJSON.deployments = deployments;
+  // check prerequisites in taskRoles
+  for (let taskRole of Object.keys(protocolJSON.taskRoles)) {
+    for (let field of prerequisiteFields) {
+      if (field in protocolJSON.taskRoles[taskRole] &&
+        !(protocolJSON.taskRoles[taskRole][field] in prerequisites[field.toLowerCase()])) {
+        throw createError(
+          'Bad Request',
+          'InvalidProtocolError',
+          `Prerequisite ${protocolJSON.taskRoles[taskRole][field]} does not exist.`
+        );
+      }
+    }
+  }
+  // check deployment in defaults
+  if ('defaults' in protocolJSON) {
+    if ('deployment' in protocolJSON.defaults &&
+      !(protocolJSON.defaults.deployment in deployments)) {
+        throw createError(
+          'Bad Request',
+          'InvalidProtocolError',
+          `Default deployment ${protocolJSON.defaults.deployment} does not exist.`
+        );
+    }
+  }
+  req.body.protocol = protocolJSON;
+  await next();
+};
+
+const protocolRender = async (req, res, next) => {
+  const protocolJSON = req.body.protocol;
+  let deployment = null;
+  if ('defaults' in protocolJSON && 'deployment' in protocolJSON.defaults) {
+    deployment = protocolJSON.deployments[protocolJSON.defaults.deployment];
+  }
+  for (let taskRole of Object.keys(protocolJSON.taskRoles)) {
+    let commands = protocolJSON.taskRoles[taskRole].commands;
+    if (deployment != null && taskRole in deployment.taskRoles) {
+      if ('preCommands' in deployment.taskRoles[taskRole]) {
+        commands = deployment.taskRoles[taskRole].preCommands.concat(commands);
+      }
+      if ('postCommands' in deployment.taskRoles[taskRole]) {
+        commands = commands.concat(deployment.taskRoles[taskRole].postCommands);
+      }
+    }
+    let entrypoint = '';
+    const tokens = mustache.parse(commands.join(' ; '), ['<%', '%>']);
```

Comment:
A simple question, whether we should use **;** or **&&** to combine commands? In **&&** case, the succeeding commands will be executed only if previous commands return an exit status of zero, this may help to find early bugs

Reply:
Update to `&&`.